IDE documentation/stubs, updated and added examples mainly for Windows usage.

TheTechsTech opened the pull request:
- added documentation for object return types with interfaces, functions, constants
  - Docblocks from readme, http://docs.libuv.org/en/v1.x/index.html, and https://github.com/thlorenz/libuv-dox
- added additional examples for using `signals` under windows
- corrected example for windows usage, to get expected results while using `uv_spawn`
- added example `poll_sendfile.php`, which reveal some functionality under windows not working,
    - the functions `uv_fs_sendfile`, `uv_fs_write`, and `uv_write` sending/writing to any uv `stream/socket/fd` client resource fails in `pipe_bind_connect.php`, `pipe.php`, `poll.php`, and `tcp_bind.php`

TheTechsTech committed:
PHP IDE documentation/stubs - WIP

Docblocks from readme, http://docs.libuv.org/en/v1.x/index.html, and https://github.com/thlorenz/libuv-dox

bwoebi commented:
Hey,

cool that you're working on it - see it's WIP, but just a quick note at a first glance:

https://github.com/bwoebi/php-uv/blob/master/php_uv.c#L2626-L2676 is defining a lot of UV* named classes. There are no resourced returned by uv_* functions, only properly typed objects.

And the functions are operating on these objects then. (You can compare that with the code, where it's using UV_PARAM_OBJ() with the appropriate class reference.
These objects btw. have no methods.

The README is a little outdated in that regard...
But these stubs are cool and we can keep them up to date; much better than that unwieldy README.

Thanks for working on it!

CharlotteDunois commented:
FYI: There are already uv stubs in [PhpStorm](https://github.com/JetBrains/phpstorm-stubs/tree/master/uv). However these are wrong, since they say the functions return resources.

cc @WyriHaximus

techno-express commented:
@bwoebi, As things currently stand, the **readme** show `resources` begin returned for all functions, which should be typed objects. The real reason for WIP, it need more interface like stubs for them, and i'm still testing out, seeing all this library functionality under windows. This extension much better than [Event](https://www.php.net/manual/en/book.event.php) and [Ev](https://www.php.net/manual/en/book.ev.php). It can even handle what [Parallel](https://www.php.net/manual/en/book.parallel.php) offer.

@CharlotteDunois, My searching did not reveal **PhpStorm** stubs, otherwise might have not even started this, anyway it does not really document or as informative as what's listed on libuv.org website.

WyriHaximus commented:
@CharlotteDunois FYI I took the readme of this extension as the source for those stubs

CharlotteDunois commented:
@WyriHaximus That's what I figured. But that doesn't make it correct. :D

@techno-express You should not think that libuv is a replacement for parallel. While libuv offers queueing work for a separate thread, there is no special handling being taken care of. I would noz be surprised if specific things would simply segfault. Parallel is explicitely made for parallelism in PHP and takes special care of everything. If you need parallelism in PHP, use parallel and not the threading work API of libuv.

techno-express commented:
@CharlotteDunois, maybe you can add some long overdue examples to [Parallel](https://github.com/krakjoe/parallel) repo.

From my programming background/schooling, I look at everything from an computer science perspective. It's just the concepts to me, and whether does it produce the results.

What is Parallel trying to achieve, how is it accomplish, how much effort is needed to get there, when is it needed? 

Anyway, I did leave some comments there, https://github.com/krakjoe/parallel/issues/59, https://github.com/krakjoe/parallel/issues/25.

I read **Parallel** documentation, it's more wrapped up in the terminology from what I see, and that can be simply matched to what's presented in **libuv**. There are workarounds to everything, it's just the developers responsibility to find it.

Take the [uv_spawn()](https://github.com/bwoebi/php-uv/blob/master/examples/spawn.php) example rewrite it using Parallel, much more work.

Parallel or parallelism have very limited uses, and can currently be accomplished using no external or third party extension.

WyriHaximus commented:
@CharlotteDunois agreed, I trusted the readme here to be correct. Maybe we can generate the stubs and readme from the ext source code to make them correct

TheTechsTech committed:
create/add interface stubs, with correct return type and parameter type hints

TheTechsTech committed:
corrections, have the return types as indicated by this library

TheTechsTech committed:
corrections, update some examples for proper operations under windows, psr-2 fixes

The `poll.php` example errors with PHP Warning:  fclose() expects parameter 1 to be resource, boolean given

The `tcp_bind.php` example does not work as expected, waits for a separate connect then does first parts and exit

The `spawn.php` example has the passed in environment array off by 1 on display

TheTechsTech committed:
Completed - added/merged in phpstorm stubs, but still needs updating with libuv docs

from https://github.com/JetBrains/phpstorm-stubs/blob/master/uv/uv_functions.php

TheTechsTech committed:
more doc updates of types, add example of sendfile on server poll, fails on windows

the `pipe.php` example errors with PHP Warning:  uv_pipe_open(): invalid argument in

It seems all examples network relate stuff fails on windows. uv_fs_write, uv_write, both return false and negative status number

TheTechsTech committed:
corrections, adding more examples of spawn with signal/process_kill, mainly for windows

TheTechsTech committed:
Update, some POSIX signals really working on Windows!

TheTechsTech committed:
Add signal constants, mainly for Windows

Modified from https://github.com/Seldaek/signal-handler/blob/master/src/SignalHandler.php

TheTechsTech committed:
some stub updates, interface, functions and types

TheTechsTech committed:
add some quick install docs

TheTechsTech committed:
Update README.md

bwoebi commented:
I've updated the inline proto stubs within php_uv.c: e76ac0dccb7908181ee715326e4f8061b77712f6

These should reflect the current state accurately.

TheTechsTech committed:
Merge remote-tracking branch 'upstream/master'

TheTechsTech committed:
Update UVFunctions.php

TheTechsTech committed:
Update UVFunctions.php

TheTechsTech committed:
Merge remote-tracking branch 'upstream/master'

TheTechsTech committed:
Merge remote-tracking branch 'upstream/master'

techno-express commented:
I see, I believed I already had those types added. Look over whenever, I think i have most things covered. 

Unless you have some corrections I need to be make, I'm mostly finish, some stuff mostly Windows I wanted to actually tests to report on.

TheTechsTech committed:
Merge remote-tracking branch 'upstream/master'

bwoebi commented:
At a quick glance it's not too bad, but there are still more than 10 signatures which still need improvement. I can check them out though in the next week.

techno-express commented:
> but there are still more than 10 signatures which still need improvement.

You right, i pull in the stub's from from https://github.com/JetBrains/phpstorm-stubs/blob/master/uv/uv_functions.php.

I sectioned them off, now realize i didn't go thru all of them. I been going thru, update and move from the section. I will finish those.

TheTechsTech committed:
Some updates of types to match current release/commits

TheTechsTech committed:
Update UVFunctions.php

TheTechsTech committed:
more type updates/corrections, and move docblocks around

TheTechsTech committed:
Merge remote-tracking branch 'upstream/master'

TheTechsTech committed:
more type updates/corrections

bwoebi committed:
UV handles are final classes, not interfaces; added missing UVLock

TheTechsTech committed:
Merge remote-tracking branch 'upstream/master'

TheTechsTech committed:
more type updates/corrections

bwoebi commented on stub/UVFunctions.php:
```diff
@@ -1840,6 +1819,97 @@ function uv_fs_event_init(UVLoop $loop, string $path, callable $callback, int $f
+ *
+ * @return void
+ */
+function ares_gethostbyname($handle, string $name, int $flag, callable $callback)
```
Where do you get these areas functions from?

techno-express replied:
Not sure now, i thought from readme or from stub i pulled in from phpstorm, i will remove.

TheTechsTech committed:
Completed docblocks, and type updates

bwoebi commented:
I think we should remove the declarations from README.md completely (in favor of the stubs)

TheTechsTech committed:
Update UVInterfaces.php

TheTechsTech committed:
Merge remote-tracking branch 'upstream/master'

bwoebi committed:
There are no methods on UV class; marked abstract

techno-express commented:
> README.md completely (in favor of the stubs)

Will do shortly.

TheTechsTech committed:
Update README.md

TheTechsTech committed:
Update UVFunctions.php

bwoebi committed:
Update UVFunctions.php

bwoebi commented on stub/UVFunctions.php:
```diff
@@ -0,0 +1,2007 @@
+ *
+ * @return int  should return UV::IS_UV_* constatns. e.g) UV::IS_UV_TCP.
+ */
+function uv_handle_type($uv)
```
It's called uv_handle_get_type in libuv. Not sure why the names are different.  ... Maybe I should rename it?

techno-express replied:
yes, i think so, it seems it returns the same thing.

TheTechsTech committed:
update rename `uv_handle_type` to `uv_handle_get_type`

TheTechsTech committed:
add uv_fs_scandir

TheTechsTech committed:
Update README.md

bwoebi commented on stub/UVFunctions.php:
```diff
@@ -0,0 +1,2039 @@
+
+/**
+ * Returns the initialized default loop.
+ * It may return NULL in case of allocation failure. This function is just a convenient
```
Maybe on the C level, but PHP will fatal error before that, Just remove that comment about alloc failure.

techno-express replied:
Will do. 

Also there many other places if viewed, will pop up with a stright line thru the comments that really doesn't to this PHP version. I will go thru and remove those too.

bwoebi commented on stub/UVFunctions.php:
```diff
@@ -0,0 +1,2039 @@
+ *
+ * @param UVPoll $poll
+ * @param int $events UV::READABLE and UV::WRITABLE flags.
+ * @param callable $callback expects (UVPoll $poll, int $status, int $events, mixed $connection)
```
mixed? can we use the actual union?

techno-express replied:
yes, and i see you have `resource $fd` in proto, will fixed.

TheTechsTech committed:
various corrections

TheTechsTech committed:
Update UVFunctions.php

TheTechsTech committed:
Update UVFunctions.php

TheTechsTech committed:
Merge remote-tracking branch 'upstream/master'

TheTechsTech committed:
add missing functions

TheTechsTech committed:
Update README.md

TheTechsTech committed:
corrections and file rename

TheTechsTech committed:
Update UVFunctions.php

TheTechsTech committed:
add missing function

TheTechsTech committed:
callback signature fixes

TheTechsTech committed:
updates, add missing handle

TheTechsTech committed:
fix `uv_fs_scandir` call signature

TheTechsTech committed:
corrections - resource to handle

TheTechsTech committed:
Update UVFunctions.php

TheTechsTech committed:
add link to doc stubs

TheTechsTech committed:
Update UVFunctions.php

TheTechsTech committed:
Update UV.php

TheTechsTech committed:
update `uv_fs_event_init` to indicate the behavior combines `uv_fs_event_start`

TheTechsTech committed:
corrections for uv_timer_start

TheTechsTech committed:
doc update from https://github.com/libuv/libuv/pull/2803

TheTechsTech committed:
indicate network socket/stream tests should be skipped as broken on Windows

https://github.com/bwoebi/php-uv/issues/78

TheTechsTech committed:
Update UV.php

TheTechsTech committed:
Update UVFunctions.php

TheTechsTech committed:
Merge remote-tracking branch 'upstream/master'

glickel commented:
Hi, is there any news ? If you need help to finish this pull, it would be a pleasure

techno-express commented:
I figure it's mostly done, when started it was focused on the previous ext-uv version, might need double checking on some callback signature changes.

glickel commented:
@techno-express so you need some times to finish this pull request ? Or do you wait for @bwoebi  to review ?

techno-express commented:
No, the PR is finish, needs review and not up to me to merge.

TheTechsTech committed:
Merge remote-tracking branch 'upstream/master'

TheTechsTech committed:
Merge remote-tracking branch 'upstream/master'

TheTechsTech committed:
Update README.md

TheTechsTech committed:
Merge branch 'master' of https://github.com/techno-express/php-uv

TheTechsTech committed:
Merge remote-tracking branch 'upstream/master'

TheTechsTech committed:
allow auto embed by build process

TheTechsTech committed:
Merge remote-tracking branch 'upstream/master'

TheTechsTech committed:
fix php_uv_address_to_zval update tests & examples

- corrections in reference to https://github.com/amphp/ext-uv/issues/78
- `uv_tcp_getsockname` returning ip4 for ip6 address, reason `uv_inet_ntop()` has `AF_INET` not `AF_INET6`

TheTechsTech committed:
Update 600-pipe_bind.phpt

TheTechsTech committed:
more `uv_queue_work()` examples

added the no real issue https://github.com/amphp/ext-uv/issues/46
